Added Active Storage support for byte ranges

tomprats opened the pull request:
### Summary

This PR allows Active Storage to have byte-range support.

I’m serving audio for a podcast using Active Storage’s proxy feature with AWS’s S3. But the audio on iOS devices weren't playing because Apple appears to (sometimes?) require byte range support. I ended up forking Rails and have been using this branch in production for a month and all seems well.

### Other Information

I opened a discussion on the forum/email list and didn't receive a lot of feedback.

https://discuss.rubyonrails.org/t/byte-range-support/76791

Let me know if there's anything else I can do to help get this moving!

santib commented on activestorage/app/controllers/active_storage/blobs/proxy_controller.rb:
```diff
@@ -6,9 +6,14 @@ class ActiveStorage::Blobs::ProxyController < ActiveStorage::BaseController
-      stream @blob
+      stream_chunk @blob
+    else
+      http_cache_forever public: true do
```
Considering the following scenario:
- First there is a request without `HTTP_RANGE` gets cached in the browser and a CDN
- Then there is a request with `HTTP_RANGE` for the same resource.

Will the browser and CDN know that they should hit the server or will they try to serve a cached response because of the `http_cache_forever public: true`?

tomprats replied:
I was trying to look into this but I'm a bit out of my depth. I can't find anything on how the browser will respond. I tried testing myself through safari and I either couldn't get it to cache the request, or it ignored the cache. Either way it still seemed to send out the byte range requests

santib commented on activestorage/app/controllers/active_storage/base_controller.rb:
```diff
@@ -9,6 +9,17 @@ class ActiveStorage::BaseController < ActionController::Base
+      chunk = blob.download_chunk range
+
+      response.header["Content-Range"] = "bytes #{range.begin}-#{range.end}/#{blob.byte_size}"
+      response.status = 206
```
shouldn't we consider the error case responding `416` status code? https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests#partial_request_responses

tomprats replied:
I initially thought this would be pretty hard to support since we'd really be dealing with how each underlying service handles the response.

Looking at it further, it looks like `Rack::Utils.get_byte_ranges` would return something like `[nil]` if there's an invalid range. I could add a check that and short circuit the it there. Something like:

```
def stream_chunk(blob)
  range = Rack::Utils.get_byte_ranges(request.headers["HTTP_RANGE"], blob.byte_size).first

  if range
    chunk = blob.download_chunk range

    response.header["Content-Range"] = "bytes #{range.begin}-#{range.end}/#{blob.byte_size}"
    response.status = 206
    response.stream.write chunk
  else
    response.status = 416
  end
ensure
  response.stream.close
end
```

santib replied:
Nice! I think that makes sense

ChrisHampel commented on activestorage/app/controllers/active_storage/base_controller.rb:
```diff
@@ -9,6 +9,17 @@ class ActiveStorage::BaseController < ActionController::Base
 
   private
+    def stream_chunk(blob)
+      range = Rack::Utils.get_byte_ranges(request.headers["HTTP_RANGE"], blob.byte_size).first
```
have you tested what happens if the client requests more than one range? I am not sure how common this is, but it is allowed by the standards, and I am not sure how clients will behave if you return only the first range they requested.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests#multipart_ranges

arpu commented:
using this branch on my cloudflare -> nginx -> puma -> rails stack and it works fine for ios safari 
only problem is cloudflare  bypass caching for mp4 per Page Rule worked

arpu commented:
response from cloudflare is 
The following have to be true:

    The origin must be sending Content-Length headers.
    The origin must be sending Accept-Range headers.

would it be possible to add the  Content-Length headers too?

arpu commented:
https://github.com/arpu/rails/commit/9f9d4578b86707eccc04a70b72ba9ecbb1a187af#diff-918b003cb18a3d542bc2fa514cde6b89569ffcea0ad1de0d6932a85a68ee4fa4

arpu commented:
can confirm this works fine with cloudflare now

arpu commented:
@tomprats should i create a new PR for this?

tomprats commented:
> @tomprats should i create a new PR for this?

Sorry I haven't been able to take another look at this in a couple weeks. I'll try to update it this week

tomprats commented:
I've updated the branch to add support a couple more things:

- Multipart ranges - Sends multiple parts separated by headers and a boundary
- Content length - Defaults to full blob's size. When it's a single range, it gives that chunk, when it's a multiple range it gives the total of all the parts sent
- Invalid ranges - When it's only 1 range it's supposed to error to send 416. When there are valid ranges with it, it simply skips the invalid

There are a couple spots I'm not confident on:
- Content length of multipart
- Boundary - I used one fo the examples from the docs but I believe we could be randomly generating it

I won't be able to test this for a little while, but would love some feedback or help. I was planning to rebase `main` back in once this was finished but let me know if you'd prefer a rebase or merge before then

arpu commented:
any Plans to merge this?

tomprats commented:
I updated this from scratch due to some restructuring and changes in ActiveStorage (original commits still available in https://github.com/tomprats/rails/tree/active-storage-byte-range-backup).

Some extra changes since last time:
- Updated boundary to be generated by SecureRandom
- Added line breaks to multipart responses
- Added some new test cases for the different types of byte range requests

Some info about the boundary: I used SecureRandom because that's what's used to generate random IDs elsewhere in Rails. RFC says it can be no longer than 70 characters, but it seemed like smaller ones were often used in the examples - https://tools.ietf.org/html/rfc2046

Here are some of the other resources if anyone wants to verify what I've been going off of:
- https://tools.ietf.org/html/rfc7233#section-4.1
- https://tools.ietf.org/html/rfc2046#section-5.1
- https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/416

Let me know if there's anything else holding it back! I'm testing it out in a couple of projects

tomprats commented:
Learned some new things tonight. Since we're streaming, we don't need the content length (we'd either have to guess at what it might be, or make all the requests first which would negate the benefits of streaming).

I'm not sure if it's not supported by curl but I can't seem to reproduce a successful multipart curl with streaming.

I was able to produce the correct output when switching from `send_stream` to `send_data`. Due to byte ranges already being smaller in size, I think it would probably be beneficial to switch to `send_data`. I will give this another look tomorrow and potentially update the PR with that solution.

tomprats commented:
I'm trying to figure out a potential issue with content-length.

1. Looks like it might be getting removed still due to something magical (does not show up in curl and produces a warning in the server's terminal).
2. The value of the header in the test does not correspond to what I set it in the proxy controller.

`send_data` might technically still be streaming, but differently than `send_stream`. Or rack or rails might be doing something I can't figure out. Something like [this](https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/actionpack/lib/action_controller/metal/streaming.rb#L206) but then I don't know why it's still in the test.

Here's a warning the server spits out when I curl the proxy. I cannot find the origin of what is producing the warning:

> [2021-04-22 23:41:14] WARN  Could not determine content-length of response body. Set content-length of the response or set Response#chunked = true

All in all, I think I've done the best I can and maybe these content length issues are meaningless. It appears to work in my app. I think I'm done until I receive some feedback. Thanks for coming to my TED talk.

tomprats commented:
Could someone take a look at this? Should be ready

cc @zzak because it looks like you've been reviewing some of the most recent active storage PRs

zzak commented:
@santib Since you have already taken a look at this wdyt?

santib commented on activestorage/app/controllers/active_storage/blobs/proxy_controller.rb:
```diff
@@ -5,8 +5,17 @@ class ActiveStorage::Blobs::ProxyController < ActiveStorage::BaseController
   include ActiveStorage::SetBlob
 
   def show
```
Just to understand the reasoning behind it. Why is this one the only default controller changed? I mean, there are 4 controllers for serving files/variations:
`activestorage/app/controllers/active_storage/blobs/proxy_controller.rb`
`activestorage/app/controllers/active_storage/blobs/redirect_controller.rb`
`activestorage/app/controllers/active_storage/representations/proxy_controller.rb`
`activestorage/app/controllers/active_storage/representations/redirect_controller.rb`

I'm not sure if keep adding logic to default controllers or if it'd be possible to achieve this in some other way. For example, shouldn't app users who want this feature create a new route + controller to implement the feature? Or why do you think this should be available(working) for everyone by default?

tomprats replied:
That's a fair point, thanks for taking a look!

I think byte range support is common enough that is almost assumed to already be supported. Each of the underlying services even utilize it through the `download_chunk` method. So with that it'd make sense to bake it into the proxy controllers.

With the redirect controllers, I am a lot less familiar. I think it is already supported because we'd pass back the url and the client would still pass the byte range header to the redirected url.

If all that makes sense to you, I will update `activestorage/app/controllers/active_storage/representations/proxy_controller.rb` to also support the byte range requests

santib replied:
>With the redirect controllers, I am a lot less familiar. I think it is already supported because we'd pass back the url and the client would still pass the byte range header to the redirected url.

In this case, wouldn't it make sense to keep the proxy controllers simpler and use the redirect urls to implement this feature in an app?

>If all that makes sense to you, I will update activestorage/app/controllers/active_storage/representations/proxy_controller.rb to also support the byte range requests

I think it makes sense to have the representations proxy controller match the blobs proxy controller behavior and maybe it'd be cool to have it extracted into a controller concern. But, just to set clear expectations, I don't have a say on what can be merged or not in Rails 😄 I'm giving feedback just to try to help PRs move forward, but then it's up to some Rails core/committer member to decide whether this is a good idea or not.

tomprats replied:
Sorry it's taken so long to get back to you, I wanted to dive into the code a little bit so I can answer better.

> In this case, wouldn't it make sense to keep the proxy controllers simpler and use the redirect urls to implement this feature in an app?

I think it'd be good to support it as much as possible. I don't think someone should have to use redirects just to get this support if proxy supports their use case better for everything else. 

> I think it makes sense to have the representations proxy controller match the blobs proxy controller behavior and maybe it'd be cool to have it extracted into a controller concern.

Totally agree with this. Most of it is in a concern already! Tonight I started updating representations to use the concern, and it was super simple until... It looks like we don't have the `byte_size` for the variants. Without the size we can't tell if the byte ranges are valid or pass the proper headers back. I'm kinda stumped here since I don't quite know how/where we could save them

> But, just to set clear expectations, I don't have a say on what can be merged or not in Rails 😄 I'm giving feedback just to try to help PRs move forward, but then it's up to some Rails core/committer member to decide whether this is a good idea or not.

I appreciate you doing this!

zzak commented:
@tomprats Could you rebase -> add a changelog -> and squash please? :bow:

tomprats commented:
@zzak Should be all set!

arpu commented:
any hope to get this in for Rails 7.0 ?

rafaelfranca commented on activestorage/app/controllers/active_storage/blobs/proxy_controller.rb:
```diff
@@ -5,8 +5,17 @@ class ActiveStorage::Blobs::ProxyController < ActiveStorage::BaseController
   def show
-    http_cache_forever public: true do
-      send_blob_stream @blob
+    if request.headers["HTTP_RANGE"].present?
```
Doesn't this work?

```suggestion
    if request.headers["Range"].present?
```

rafaelfranca commented on activestorage/app/controllers/active_storage/blobs/proxy_controller.rb:
```diff
@@ -5,8 +5,17 @@ class ActiveStorage::Blobs::ProxyController < ActiveStorage::BaseController
+    if request.headers["HTTP_RANGE"].present?
+      response.headers["Accept-Ranges"] = "bytes"
+
+      send_blob_byte_range_data @blob, request.headers["HTTP_RANGE"]
```
```suggestion
      send_blob_byte_range_data @blob, request.headers["Range"]
```

rafaelfranca commented on activestorage/app/controllers/active_storage/blobs/proxy_controller.rb:
```diff
@@ -5,8 +5,17 @@ class ActiveStorage::Blobs::ProxyController < ActiveStorage::BaseController
-    http_cache_forever public: true do
-      send_blob_stream @blob
+    if request.headers["HTTP_RANGE"].present?
+      response.headers["Accept-Ranges"] = "bytes"
```
Is there any reason why this is being set here and not in the `send_blob_byte_range_data`?

tomprats replied:
I left it out to keep it in the same place as the one on the other side of the conditional, but it probably does make more sense in that method since if you're sending byte range data you definitely accept byte ranges.

I'll move it

rafaelfranca commented on activestorage/app/controllers/concerns/active_storage/streaming.rb:
```diff
@@ -1,11 +1,54 @@
+        data,
+        disposition: blob.forced_disposition_for_serving || disposition || DEFAULT_BLOB_STREAMING_DISPOSITION,
+        filename: blob.filename.sanitized,
+        status: 206,
```
```suggestion
        status: :partial_content,
```

rafaelfranca commented on activestorage/app/controllers/concerns/active_storage/streaming.rb:
```diff
@@ -1,11 +1,54 @@
+      else
+        boundary = SecureRandom.hex
+        content_type = "multipart/byteranges; boundary=#{boundary}"
+        data = ""
```
```suggestion
        data = +""
```

rafaelfranca commented on activestorage/app/controllers/concerns/active_storage/streaming.rb:
```diff
@@ -1,11 +1,54 @@
+        ranges.compact.each do |range|
+          chunk = blob.download_chunk(range)
+
+          data += "\r\n--#{boundary}\r\n"
```
```suggestion
          data << "\r\n--#{boundary}\r\n"
```

rafaelfranca commented on activestorage/app/controllers/concerns/active_storage/streaming.rb:
```diff
@@ -1,11 +1,54 @@
+          chunk = blob.download_chunk(range)
+
+          data += "\r\n--#{boundary}\r\n"
+          data += "Content-Type: #{blob.content_type_for_serving}\r\n"
```
```suggestion
          data << "Content-Type: #{blob.content_type_for_serving}\r\n"
```

rafaelfranca commented on activestorage/app/controllers/concerns/active_storage/streaming.rb:
```diff
@@ -1,11 +1,54 @@
+
+          data += "\r\n--#{boundary}\r\n"
+          data += "Content-Type: #{blob.content_type_for_serving}\r\n"
+          data += "Content-Range: bytes #{range.begin}-#{range.end}/#{blob.byte_size}\r\n\r\n"
```
```suggestion
          data << "Content-Range: bytes #{range.begin}-#{range.end}/#{blob.byte_size}\r\n\r\n"
```

rafaelfranca commented on activestorage/app/controllers/concerns/active_storage/streaming.rb:
```diff
@@ -1,11 +1,54 @@
+          data += "\r\n--#{boundary}\r\n"
+          data += "Content-Type: #{blob.content_type_for_serving}\r\n"
+          data += "Content-Range: bytes #{range.begin}-#{range.end}/#{blob.byte_size}\r\n\r\n"
+          data += chunk
```
```suggestion
          data << chunk
```

rafaelfranca commented on activestorage/app/controllers/concerns/active_storage/streaming.rb:
```diff
@@ -1,11 +1,54 @@
+          data += chunk
+        end
+
+        data += "\r\n--#{boundary}--\r\n"
```
```suggestion
        data << "\r\n--#{boundary}--\r\n"
```

tomprats commented:
Sorry for the wait, I was on vacation for a week. Going to take a look at this tonight or this weekend. Are commit suggestions good to use or will they create a ton of commits?

tomprats committed:
Squashed commit of the following:

commit 1dca99ad4f8082d8daaa17c6600f3036c25f8e50
Author: Tom Prats <tprats108@gmail.com>
Date:   Thu Aug 5 18:07:00 2021 -0400

    Moved header

commit 62201870ff5c4124d90912989745819b05d94516
Merge: 5fa3ecae74 c91a8135c7
Author: Tom Prats <tprats108@gmail.com>
Date:   Thu Aug 5 14:41:22 2021 -0400

    Merge branch 'main' into active-storage-byte-range

commit 5fa3ecae745b4f7c67a6b6b1b7ec420877c96fb8
Author: Tom Prats <tprats108@gmail.com>
Date:   Thu Aug 5 14:39:53 2021 -0400

    Apply suggestions from code review

    Syntax updates

    Co-authored-by: Rafael França <rafael@franca.dev>

commit b9553e3698a7af5105171f9d63bd7b89cbb7e2c3
Author: Tom Prats <tprats108@gmail.com>
Date:   Wed Jun 23 17:36:26 2021 -0400

    Added Active Storage support for byte ranges

tomprats commented:
- Updated everything with your feedback @rafaelfranca
- Re-squashed it down into 1 commit again

kaspth commented:
@tomprats perfect, thank you!

johanandre commented:
Is there any chance this would be merged to 6.1.4?

giedomak commented:
> can confirm this works fine with cloudflare now

@arpu, did you manage to cache the 206 partial responses by cloudflare? If so, could you please provide some pointers how to get this to work?

arpu commented:
@giedomak  still using 6.x  because this is not in the 7.x release ( but you can use rails master)

skipkayhil commented:
> this is not in the 7.x release

If you look at the merged commit https://github.com/rails/rails/commit/fe4ec2ac0c131013090d9f36bceb3ce4a084af82 it shows that this has been in every release since 7.0.0.alpha1

arpu commented:
@skipkayhil  ah yes you are right but this has a bug and it fixed in https://github.com/rails/rails/pull/45102
